File: 06-TREES/kth-smallest-element-in-a-bst.py
from typing import *
from TreeNode import TreeNode


# https://leetcode.com/problems/kth-smallest-element-in-a-bst/
class Solution:
    def __init__(self):
        self.inordered = []

    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        if root:
            self.kthSmallest(root.left, k)

            if len(self.inordered) < k:
                self.inordered.append(root.val)
            else:
                return self.inordered[k-1]

            self.kthSmallest(root.right, k)

        if len(self.inordered) >= k:
            return self.inordered[k - 1]


root = TreeNode.fromArray([5, 3, 6, 2, 4, None, None, 1])
root.print()
print(Solution().kthSmallest(root, 3))

File: 06-TREES/kth-smallest-element-in-a-bst.php
<?php

# https://leetcode.com/problems/kth-smallest-element-in-a-bst/
class Solution
{
    private array $inordered = [];

    public function kthSmallest(?TreeNode $root, int $k): ?int
    {
        if ($root) {
            $this->kthSmallest($root->left, $k);

            if (count($this->inordered) < $k)
                $this->inordered[] = $root->val;
            else
                return $this->inordered[$k - 1];

            $this->kthSmallest($root->right, $k);
        }

        return $this->inordered[$k - 1] ?? null;
    }
}

